ring/ownership_priority_queue.go

```diff
@@ -0,0 +1,145 @@
+package ring
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+type ringItem interface {
+	key() int
+	String() string
+}
+
+type ringInstance struct {
+	instanceID int
+}
+
+func (ri ringInstance) key() int {
+	return ri.instanceID
+}
+
+func (ri ringInstance) String() string {
+	return fmt.Sprintf("[instanceID: %d]", ri.instanceID)
+}
+
+type ringToken struct {
+	token     uint32
+	prevToken uint32
+}
+
+func (rt ringToken) key() int {
+	return int(rt.token)
+}
+
+func (rt ringToken) String() string {
+	return fmt.Sprintf("[token: %d, prevToken: %d]", rt.token, rt.prevToken)
+}
+
+type ownershipInfo[T ringItem] struct {
+	item      T
+	ownership float64
+}
+
+func newRingTokenOwnershipInfo(token, prevToken uint32) ownershipInfo[ringToken] {
+	ownership := float64(getTokenDistance(prevToken, token))
+	return ownershipInfo[ringToken]{
+		ownership: ownership,
+		item: ringToken{
+			token:     token,
+			prevToken: prevToken,
+		},
+	}
+}
+
+func newRingInstanceOwnershipInfo(instanceID int, ownership float64) ownershipInfo[ringInstance] {
+	return ownershipInfo[ringInstance]{
+		ownership: ownership,
+		item: ringInstance{
+			instanceID: instanceID,
+		},
+	}
+}
+
+// ownershipPriorityQueue is a max-heap, i.e., a priority queue
+// where items with a higher priority will be extracted first.
+// Namely, items with a higher ownership have a higher priority.
+// In order to guarantee that 2 instances of ownershipPriorityQueue
+// with the same items always assign equal priorities to equal items,
+// in the case of items with equal ownership, we rely on the
+// order of item ids.
+type ownershipPriorityQueue[T ringItem] struct {
+	items []ownershipInfo[T]
+}
+
+func newPriorityQueue[T ringItem](len int) ownershipPriorityQueue[T] {
+	return ownershipPriorityQueue[T]{
+		items: make([]ownershipInfo[T], 0, len),
```

Comment:
Let's reduce possible confusions, since we're not building a slice of that length, but that capacity.
```suggestion
func newPriorityQueue[T ringItem](capacity int) ownershipPriorityQueue[T] {
	return ownershipPriorityQueue[T]{
		items: make([]ownershipInfo[T], 0, capacity),
```